python/tskit/combinatorics.py

```diff
@@ -0,0 +1,352 @@
+
+    n = tree.num_samples()
+    for _ in range(n):
+        tables.nodes.add_row(flags=tskit.NODE_IS_SAMPLE, time=0)
+    _label_unrank(tree.root, label_rank, list(range(n)), n)
+
+    tables.sort()
+    return tables.tree_sequence().first()
+
+
+def S(n):
+    """
+    The cardinality of the set of unlabeled binary trees with n leaves,
+    under the conditions of tree isomorphism universal to this module.
+    """
+    if n <= 1:
+        return n
+
+    total = 0
+    for k in range(math.ceil(n / 2), n):
+        if k == n // 2:
+            total += (S(k) * (S(k) + 1)) // 2
+        else:
+            total += S(k) * S(n - k)
+    return total
+
+
+def L(shape_rank, n):
+    if n == 1:
+        assert shape_rank == 0
+        return 1
+
+    k, l_rank, r_rank = subtree_shape_ranks(shape_rank, n)
+    total = nCr(n, k) * L(l_rank, k) * L(r_rank, n - k)
+
+    # If the tree is symmetric, there is half the number of labelings
+    if k == n - k and l_rank == r_rank:
+        assert total % 2 == 0
+        return total // 2
+    return total
+
+
+def subtree_shape_ranks(rank, n):
+    assert n > 1
+    assert 0 <= rank < S(n), f"{n} | {S(n)} | {rank}"
+    x = 0
+    k = n - 1
+    while x + S(k) * S(n - k) <= rank:
+        x += S(k) * S(n - k)
+        k -= 1
+    assert k >= math.ceil(n / 2)
+    rank_k = rank - x
+
+    if k == n - k:
+        l_rank = 0
+        y = 0
+        # TODO Probably a closed form solution to this
+        while y + S(k) - l_rank <= rank_k:
+            y += S(k) - l_rank
+            l_rank += 1
+        r_rank = rank_k - y + l_rank
+    else:
+        l_rank = rank_k // S(n - k)
+        r_rank = rank_k % S(n - k)
+
+    assert 0 <= l_rank < S(k)
+    assert 0 <= r_rank < S(n - k)
+
+    return k, l_rank, r_rank
+
+
+def L_tree(tree, u=None):
+    if u is None:
+        u = tree.root
+
+    if tree.is_leaf(u):
+        return 1
+
+    left, right = canon_left_right(tree, u)
+    n = tree.num_samples(u)
+    k = tree.num_samples(left)
+    total = nCr(n, k) * L_tree(tree, left) * L_tree(tree, right)
+
+    if is_symmetrical(tree, u):
+        return total / 2
+    return total
+
+
+def is_symmetrical(tree, u):
+    left = tree.left_child(u)
+    right = tree.right_child(u)
+    return tree.num_samples(left) == tree.num_samples(right) and shape_rank(
+        tree, left
+    ) == shape_rank(tree, right)
+
+
+def is_canonical(tree, u=None):
+    """
+    Checks for the canonical assignment of labels on a tree. We
+    don't care/enforce that tskit trees be canonically oriented with
+    left/right, because edge orders are governed by time, etc.
+    Instead, we check that under a canonical orientation, the labeling is
+    canonical.
+    """
+    if u is None:
+        u = tree.root
+
+    if tree.is_leaf(u):
+        return True
+
+    left, right = canon_left_right(tree, u)
+    if is_symmetrical(tree, u):
+        return min(tree.samples(left)) < min(tree.samples(right))
+
+    return True
+
+
+def nCr(n, r):
```

Comment:
Huh, annoying that numpy doesn't provide a definition of this. Looks like python 3.8+ has the [comb](https://docs.python.org/3.8/library/math.html?highlight=comb#math.comb) function.

Perhaps we should follow its API, and look to drop our own definition at point in the future?

Reply:
my thoughts as well

Reply:
my thoughts as well